### PR TITLE
New Rule - no-duplicate-hooks

### LINKED ISSUE #231: New rule: no duplicated hooks
A rule that disallows duplicated hooks in the same describe.
For example this code will fail, because you should combine the two `beforeEach`s into one:

```
describe('my describe', () => {
    beforeEach(() => {...});
    beforeEach(() => {...});

    it('my it', () => {...})
});
```

### PR DESCRIPTION
**Pull request by M4rk9696**
Fix #231

### CODE REVIEW

File: src/rules/util.js [hunk 1]
```diff
@@ -99,6 +99,13 @@ const testCaseNames = Object.assign(Object.create(null), {
   xtest: true,
 });
 
+const testHookNames = Object.assign(Object.create(null), {
```
Comment (jeysal) [inline src/rules/util.js]: Can we use a Set for this? (and change the code above as well)
This way of doing it is probably from before Sets were supported everywhere.

File: src/rules/__tests__/no-duplicate-hooks.test.js [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+const rule = require('../no-duplicate-hooks');
+
+const ruleTester = new RuleTester({
+  parserOptions: {
+    ecmaVersion: 6,
+  },
+});
+
+ruleTester.run('basic describe block', rule, {
+  valid: [
```
Comment (jeysal) [inline src/rules/__tests__/no-duplicate-hooks.test.js]: I think an additional test checking that different kinds of hooks in one describe block are allowed would be good

File: src/rules/util.js (hunk 1 above)
Comment (M4rk9696) [reply src/rules/util.js]: I thought it was the standard followed in the project, will change it

Comment (jeysal) [reply src/rules/util.js]: It probably was so far, but I imagine only for historical reasons so it's a good time to change that when touching this code :)

File: src/rules/no-focused-tests.js [hunk 3]
```diff
@@ -2,16 +2,14 @@
...
 
-const testFunctions = Object.assign(Object.create(null), {
-  describe: true,
-  it: true,
-  test: true,
-});
+const testFunctions = new Set(['describe', 'it', 'test']);
 
-const matchesTestFunction = object => object && testFunctions[object.name];
+const matchesTestFunction = object => object && testFunctions.has(object.name);
```
Comment (SimenB) [inline src/rules/no-focused-tests.js]: way better, thanks!

**Pull request merged**